add static build config. (#510)
shibatch/sleef: CMakeLists.txt

# Options

> option(SLEEF_BUILD_STATIC_TEST_BINS "Build statically linked test executables" OFF)

Reviewer: The behavior of `option` and `set` are controlled by policies. You _must_ set `cmake_minimum_required` as the first line of your top-level program to have _any_ backwards compatibility guarantees at all.

Ideally `project()` would be the second line since _so many_ other commands depend on the internal state it configures.

```suggestion
cmake_minimum_required(VERSION 3.18)
project(SLEEF LANGUAGES C VERSION 3.6.0)

# TODO: rename uses of this variable to use the standard one
set(SLEEF_VERSION_PATCHLEVEL ${SLEEF_VERSION_PATCH})
set(SLEEF_SOVERSION ${SLEEF_VERSION_MAJOR})

option(SLEEF_BUILD_STATIC_TEST_BINS "Build statically linked test executables" OFF)
```

Then, after your options, add the following lines:
```
set(
if (SLEEF_ENABLE_CXX)
  enable_language(CXX)
endif ()
if (SLEEF_ENABLE_CUDA)
  enable_language(CUDA)
endif ()
```

No need to maintain that `LANGLIST` anymore. Just enable the requested languages.

Then, below, delete these lines:

```
cmake_minimum_required(VERSION 3.18)
if(${CMAKE_VERSION} VERSION_GREATER "3.14.99")
  cmake_policy(SET CMP0091 NEW)
endif()
```

`CMP0091` is enabled in CMake 3.15+ so the `cmake_policy` line is a noop and so the whole `if ()` block can be removed.

Also delete these lines:

```
set(SLEEF_VERSION_MAJOR 3)
set(SLEEF_VERSION_MINOR 6)
set(SLEEF_VERSION_PATCHLEVEL 0)
set(SLEEF_VERSION ${SLEEF_VERSION_MAJOR}.${SLEEF_VERSION_MINOR}.${SLEEF_VERSION_PATCHLEVEL})
set(SLEEF_SOVERSION ${SLEEF_VERSION_MAJOR})
set(LANGLIST C)
if (SLEEF_ENABLE_CUDA)
  set(LANGLIST ${LANGLIST} CUDA)
endif()
if (SLEEF_ENABLE_CXX)
  set(LANGLIST ${LANGLIST} CXX)
endif()
```

# Options

> option(SLEEF_BUILD_SHARED_LIBS "Build sleef as static library" ON)

Reviewer 2: Can you try this change instead and switch back to old name everywhere? I believe this can work according to the documentation (https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html)
```
-option(SLEEF_BUILD_SHARED_LIBS "Build sleef as static library" ON)
+# Declare BUILD_SHARED_LIBS option to make it accessible to top-level projects.
+option(BUILD_SHARED_LIBS "Build using shared libraries (CMake reserved)" ON)
+
```

Author: I have tried this method, It not constant work, the reason as above: https://github.com/shibatch/sleef/pull/510#issuecomment-1934336010

Author: pytorch default want setup “BUILD_SHARED_LIBS” as shared, but it want to use sleef as static library. Current pytorch is used a save/restore this variable to WA it. 
Save: https://github.com/pytorch/pytorch/blob/cee16353db384dd33a4804f9f26829cacfcfbab7/aten/src/ATen/CMakeLists.txt#L424-L440
Restore: https://github.com/pytorch/pytorch/blob/cee16353db384dd33a4804f9f26829cacfcfbab7/aten/src/ATen/CMakeLists.txt#L480-L485
And the original author think we should use SLEEF_ to pass the sleef build config also: https://github.com/pytorch/pytorch/blob/cee16353db384dd33a4804f9f26829cacfcfbab7/aten/src/ATen/CMakeLists.txt#L483

option(SLEEF_ENABLE_CUDA "Enable CUDA" OFF)
option(SLEEF_ENABLE_CXX "Enable C++" OFF)

set(_INTERNAL_SLEEF_BUILD_SHARED_LIBS)
if(NOT DEFINED ${SLEEF_BUILD_SHARED_LIBS})
  set(_INTERNAL_SLEEF_BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS})
else()
  set(_INTERNAL_SLEEF_BUILD_SHARED_LIBS ${SLEEF_BUILD_SHARED_LIBS})
endif()

> 

Reviewer 2: Can this be done with options instead? For instance, 
```
option(BUILD_SHARED_LIBS "Build sleef as shared library" ON)
option(SLEEF_BUILD_SHARED_LIBS "Build sleef as shared library" ${BUILD_SHARED_LIBS})
```

Reviewer: @blapie -- These are not equivalent because this defines `SLEEF_BUILD_SHARED_LIBS` as a cache variable when no variable by that name (normal or cache) exists. So then changing `BUILD_SHARED_LIBS` only works as expected on the first configuration.

My blog post suggests the following:

```suggestion
if (DEFINED SLEEF_BUILD_SHARED_LIBS)
  set(BUILD_SHARED_LIBS ${SLEEF_BUILD_SHARED_LIBS})
endif ()
```

Then you can just look at the truthiness of `BUILD_SHARED_LIBS` down the line. If you care about documenting this variable in the CMake CLI/GUI, then this approach works:

```suggestion
set(SLEEF_BUILD_SHARED_LIBS "undefined"
    CACHE BOOL "Override BUILD_SHARED_LIBS while configuring SLEEF")
if (NOT SLEEF_BUILD_SHARED_LIBS STREQUAL "undefined")
  set(BUILD_SHARED_LIBS ${SLEEF_BUILD_SHARED_LIBS})
endif ()

> `CMakeCache.txt` and directory `CMakeFiles` in the current directory.")

Reviewer: Everywhere you should just use `SLEEF_SOURCE_DIR` and `SLEEF_BINARY_DIR` since those are the standard and expected names. The `project` command defines them and the `VERSION` variables.

https://cmake.org/cmake/help/v3.18/command/project.html#project

endif()

> if(SLEEF_ENABLE_LTO AND _INTERNAL_SLEEF_BUILD_SHARED_LIBS)
>   message(FATAL_ERROR "SLEEF_ENABLE_LTO and _INTERNAL_SLEEF_BUILD_SHARED_LIBS cannot be specified at the same time")
> endif(SLEEF_ENABLE_LTO AND _INTERNAL_SLEEF_BUILD_SHARED_LIBS)

Reviewer: With my other suggestions:

```suggestion
if (SLEEF_ENABLE_LTO AND BUILD_SHARED_LIBS)
  message(FATAL_ERROR "SLEEF_ENABLE_LTO and (SLEEF_)BUILD_SHARED_LIBS cannot be enabled at the same time")
endif ()
```
Note that duplicating the condition in the `endif` hasn't been a requirement since at _least_ 2.8.x but it might be as far back as 2.4.x. Definitely not needed in 3.18, which is your minimum.

if(SLEEF_ENABLE_LTO)

>   cmake_policy(SET CMP0069 NEW)

Reviewer: ```suggestion
```
That policy is already enabled by `cmake_minimum_required(VERSION 3.18)`. It's enabled for 3.9+.

    message("   Native build dir: ${NATIVE_BUILD_DIR}")
  endif(CMAKE_CROSSCOMPILING)
  message(STATUS "Using option `${SLEEF_C_FLAGS}` to compile libsleef")

>   message(STATUS "Building shared libs : " ${_INTERNAL_SLEEF_BUILD_SHARED_LIBS})

Reviewer: ```suggestion
  message(STATUS "Building shared libs : " ${BUILD_SHARED_LIBS})
```
